tpu/templates/tpu_film_estimator/trainer.py

```diff
@@ -0,0 +1,266 @@
+        x_shape = x.get_shape().merge_with([batch_size, None, None, None])
+        y_shape = y.get_shape().merge_with([batch_size])
+        modulation_data_shape = modulation_data.get_shape().merge_with([batch_size, None])
+
+        x.set_shape(x_shape)
+        y.set_shape(y_shape)
+        modulation_data.set_shape(modulation_data_shape)
+
+        # Also format the dataset with a dict for features
+        features = {'x': x, 'modulation_data': modulation_data}
+        labels = y
+
+        return features, labels
+
+    dataset = dataset.map(set_shapes_and_format)
+    dataset = dataset.prefetch(tf.contrib.data.AUTOTUNE)
+
+    return dataset
+
+
+def main(args):
+    # pass the args as params so the model_fn can use
+    # the TPU specific args
+    params = vars(args)
+
+    if args.use_tpu:
+        # additional configs required for using TPUs
+        tpu_cluster_resolver = tf.contrib.cluster_resolver.TPUClusterResolver(args.tpu)
+        tpu_config = tf.contrib.tpu.TPUConfig(
+            num_shards=8, # using Cloud TPU v2-8
+            iterations_per_loop=args.save_checkpoints_steps)
+
+        # use the TPU version of RunConfig
+        config = tf.contrib.tpu.RunConfig(
+            cluster=tpu_cluster_resolver,
+            model_dir=args.model_dir,
+            tpu_config=tpu_config,
+            save_checkpoints_steps=args.save_checkpoints_steps,
+            save_summary_steps=100)
+
+        # TPUEstimator
+        estimator = tf.contrib.tpu.TPUEstimator(
+            model_fn=model_fn,
+            config=config,
+            params=params,
+            train_batch_size=args.train_batch_size,
+            eval_batch_size=32, # FIXME
+            export_to_tpu=False)
+    else:
+        config = tf.estimator.RunConfig(model_dir=args.model_dir)
+
+        estimator = tf.estimator.Estimator(
+            model_fn,
+            config=config,
+            params=params)
+
+    estimator.train(train_input_fn, max_steps=args.max_steps)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
+        '--model-dir',
+        type=str,
+        default='/tmp/tpu-template')
+    parser.add_argument(
+        '--max-steps',
+        type=int,
+        default=1000)
+    parser.add_argument(
+        '--train-batch-size',
+        type=int,
+        default=16)
+    parser.add_argument(
+        '--save-checkpoints-steps',
+        type=int,
+        default=100)
+    parser.add_argument(
+        '--use-tpu',
+        action='store_true')
+    parser.add_argument(
+        '--tpu',
+        default=None)
+
+    args, _ = parser.parse_known_args()
+
+    # colab.research.google.com specific
+    import sys
+    if 'google.colab' in sys.modules:
+        import json
+        import os
+
+        # TODO(user): change this
+        args.model_dir = 'gs://your-gcs-bucket'
+
+        # When connected to the TPU runtime
+        if 'COLAB_TPU_ADDR' in os.environ:
```

Comment:
Not at all.  The .py --> .ipynb tool is too lazy about this.  Let me fix it in a separate PR since that fix will touch several files not in the scope of this PR.

Reply:
LGTM